File: sample/src/main/java/pl/charmas/android/reactivelocation2/sample/PlacesActivity.java
package pl.charmas.android.reactivelocation2.sample;

import android.location.Location;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;

import com.google.android.gms.location.places.AutocompletePrediction;
import com.google.android.gms.location.places.AutocompletePredictionBuffer;
import com.google.android.gms.location.places.PlaceLikelihood;
import com.google.android.gms.location.places.PlaceLikelihoodBuffer;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.disposables.CompositeDisposable;
import io.reactivex.rxjava3.functions.BiFunction;
import io.reactivex.rxjava3.functions.Consumer;
import io.reactivex.rxjava3.functions.Function;
import io.reactivex.rxjava3.functions.Predicate;
import pl.charmas.android.reactivelocation2.ReactiveLocationProvider;
import pl.charmas.android.reactivelocation2.sample.utils.RxTextView;

import static pl.charmas.android.reactivelocation2.sample.utils.UnsubscribeIfPresent.dispose;

public class PlacesActivity extends BaseActivity {

    private TextView currentPlaceView;
    private EditText queryView;
    private ListView placeSuggestionsList;
    private ReactiveLocationProvider reactiveLocationProvider;
    private CompositeDisposable compositeDisposable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_places);
        currentPlaceView = (TextView) findViewById(R.id.current_place_view);
        queryView = (EditText) findViewById(R.id.place_query_view);
        placeSuggestionsList = (ListView) findViewById(R.id.place_suggestions_list);
        placeSuggestionsList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                AutocompleteInfo info = (AutocompleteInfo) parent.getAdapter().getItem(position);
                startActivity(PlacesResultActivity.getStartIntent(PlacesActivity.this, info.id));
            }
        });

        reactiveLocationProvider = new ReactiveLocationProvider(this);
    }

    @Override
    protected void onLocationPermissionGranted() {
        compositeDisposable = new CompositeDisposable();
        compositeDisposable.add(
                reactiveLocationProvider.getCurrentPlace(null)
                        .subscribe(new Consumer<PlaceLikelihoodBuffer>() {
                            @Override
                            public void accept(PlaceLikelihoodBuffer buffer) {
                                PlaceLikelihood likelihood = buffer.get(0);
                                if (likelihood != null) {
                                    currentPlaceView.setText(likelihood.getPlace().getName());
                                }
                                buffer.release();
                            }
                        }, new Consumer<Throwable>() {
                            @Override
                            public void accept(Throwable throwable) throws Exception {
                                Log.e("PlacesActivity", "Error in observable", throwable);
                            }
                        })
        );

        Observable<String> queryObservable = RxTextView
                .textChanges(queryView)
                .map(new Function<CharSequence, String>() {
                    @Override
                    public String apply(CharSequence charSequence) {
                        return charSequence.toString();
                    }
                })
                .debounce(1, TimeUnit.SECONDS)
                .filter(new Predicate<String>() {
                    @Override
                    public boolean test(String s) {
                        return !TextUtils.isEmpty(s);
                    }
                });
        Observable<Location> lastKnownLocationObservable = reactiveLocationProvider.getLastKnownLocation();
        Observable<AutocompletePredictionBuffer> suggestionsObservable = Observable
                .combineLatest(queryObservable, lastKnownLocationObservable,
                        new BiFunction<String, Location, QueryWithCurrentLocation>() {
                            @Override
                            public QueryWithCurrentLocation apply(String query, Location currentLocation) {
                                return new QueryWithCurrentLocation(query, currentLocation);
                            }
                        }).flatMap(new Function<QueryWithCurrentLocation, Observable<AutocompletePredictionBuffer>>() {
                    @Override
                    public Observable<AutocompletePredictionBuffer> apply(QueryWithCurrentLocation q) {
                        if (q.location == null) return Observable.empty();

                        double latitude = q.location.getLatitude();
                        double longitude = q.location.getLongitude();
                        LatLngBounds bounds = new LatLngBounds(
                                new LatLng(latitude - 0.05, longitude - 0.05),
                                new LatLng(latitude + 0.05, longitude + 0.05)
                        );
                        return reactiveLocationProvider.getPlaceAutocompletePredictions(q.query, bounds, null);
                    }
                });

        compositeDisposable.add(suggestionsObservable.subscribe(new Consumer<AutocompletePredictionBuffer>() {
            @Override
            public void accept(AutocompletePredictionBuffer buffer) {
                List<AutocompleteInfo> infos = new ArrayList<>();
                for (AutocompletePrediction prediction : buffer) {
                    infos.add(new AutocompleteInfo(prediction.getFullText(null).toString(), prediction.getPlaceId()));
                }
                buffer.release();
                placeSuggestionsList.setAdapter(new ArrayAdapter<>(PlacesActivity.this, android.R.layout.simple_list_item_1, infos));
            }
        }));
    }

    @Override
    protected void onStop() {
        super.onStop();
        dispose(compositeDisposable);
    }

    private static class QueryWithCurrentLocation {
        final String query;
        public final Location location;

        private QueryWithCurrentLocation(String query, Location location) {
            this.query = query;
            this.location = location;
        }
    }

    private static class AutocompleteInfo {
        private final String description;
        private final String id;

        private AutocompleteInfo(String description, String id) {
            this.description = description;
            this.id = id;
        }

        @Override
        public String toString() {
            return description;
        }
    }
}

File: sample/src/main/java/pl/charmas/android/reactivelocation2/sample/PlacesActivity.kt
package pl.charmas.android.reactivelocation2.sample

import android.annotation.SuppressLint
import android.location.Location
import android.os.Bundle
import android.util.Log
import android.widget.AdapterView.OnItemClickListener
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.ListView
import android.widget.TextView
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.RectangularBounds
import com.google.android.libraries.places.api.net.FindAutocompletePredictionsRequest
import com.google.android.libraries.places.api.net.FindCurrentPlaceRequest
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.BiFunction
import pl.charmas.android.reactivelocation2.ReactiveLocationProvider
import pl.charmas.android.reactivelocation2.sample.utils.RxTextView
import pl.charmas.android.reactivelocation2.sample.utils.UnsubscribeIfPresent
import java.util.concurrent.TimeUnit

class PlacesActivity : BaseActivity() {

    private lateinit var currentPlaceView: TextView
    private lateinit var queryView: EditText
    private lateinit var placeSuggestionsList: ListView
    private lateinit var reactiveLocationProvider: ReactiveLocationProvider
    private var compositeDisposable: CompositeDisposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_places)
        Places.initialize(this.applicationContext, getString(R.string.API_KEY))
        currentPlaceView = findViewById(R.id.current_place_view)
        queryView = findViewById(R.id.place_query_view)
        placeSuggestionsList = findViewById(R.id.place_suggestions_list)
        placeSuggestionsList.onItemClickListener =
            OnItemClickListener { parent, view, position, id ->
                val info =
                    parent.adapter.getItem(position) as AutocompleteInfo
                val placeId = info.id
                if (placeId != null) {
                    startActivity(PlacesResultActivity.getStartIntent(this@PlacesActivity, placeId))
                }
            }
        reactiveLocationProvider = ReactiveLocationProvider(this)
    }

    @SuppressLint("MissingPermission")
    override fun onLocationPermissionGranted() {
        compositeDisposable.add(
            reactiveLocationProvider.getCurrentPlace(
                FindCurrentPlaceRequest.builder(listOf(com.google.android.libraries.places.api.model.Place.Field.ID))
                    .build()
            )
                .subscribe({ response ->
                    val buffer = response.placeLikelihoods
                    val likelihood = buffer.firstOrNull()
                    if (likelihood != null) {
                        currentPlaceView.text = likelihood.place.name
                    }
                }) { throwable ->
                    Log.e("PlacesActivity", "Error in observable", throwable)
                }
        )
        val queryObservable = RxTextView
            .textChanges(queryView)
            .map { charSequence -> charSequence.toString() }
            .debounce(1, TimeUnit.SECONDS)
            .filter { s -> s.isNotEmpty() }

        compositeDisposable.add(
            Observable.combineLatest(
                queryObservable,
                reactiveLocationProvider.lastKnownLocation,
                BiFunction<String, Location, QueryWithCurrentLocation> { query, currentLocation ->
                    QueryWithCurrentLocation(query, currentLocation)
                }
            )
                .flatMapMaybe { q ->
                    val latitude = q.location.latitude
                    val longitude = q.location.longitude
                    val bounds = LatLngBounds(
                        LatLng(latitude - 50.05, longitude - 50.05),
                        LatLng(latitude + 50.05, longitude + 50.05)
                    )
                    reactiveLocationProvider.getPlaceAutocompletePredictions(
                        FindAutocompletePredictionsRequest.builder()
                            .setQuery(q.query)
                            .setLocationRestriction(RectangularBounds.newInstance(bounds))
                            .build()
                    )
                }
                .doOnError { Log.e(TAG, "onLocationPermissionGranted (line 80): ", it) }
                .retry()
                .subscribe { buffer ->
                    val infos = mutableListOf<AutocompleteInfo>()
                    for (prediction in buffer) {
                        infos.add(

                            AutocompleteInfo(
                                prediction.getFullText(null).toString(),
                                prediction.placeId
                            )
                        )
                    }
                    placeSuggestionsList.adapter = ArrayAdapter(
                        this@PlacesActivity,
                        android.R.layout.simple_list_item_1,
                        infos
                    )
                }
        )
    }

    override fun onStop() {
        super.onStop()
        UnsubscribeIfPresent.dispose(compositeDisposable)
    }

    data class QueryWithCurrentLocation(val query: String, val location: Location)

    data class AutocompleteInfo(
        val description: String,
        val id: String?
    ) {
        override fun toString(): String {
            return description
        }
    }

    companion object {
        const val TAG: String = "PlacesActivity"
    }
}